### PR TITLE
[MemoryBuiltins] Cache the result of ObjectOffsetSizeVisitor::visit. #64796

### PR DESCRIPTION
**Pull request by bevin-hansson**
visit will skip visiting instructions it already has visited
to avoid issues with cycles in the data graph. However,
the result of this skipping behavior is that if we
encounter the same instruction twice, and that instruction
has a well defined result and isn't part of a cycle, we
will introduce unknowns into the analysis even though we
knew the size and offset of the instruction's result.

Instead of skipping such instructions, keep a cache of
the result of visiting them. This result is initialized
to unknown() before visiting, so if we happen to visit
it again recursively (perhaps as the result of a cycle
or a phi), we will get unknown as the cached result and
exit out.

### CODE REVIEW

File: llvm/test/Transforms/LowerConstantIntrinsics/builtin-object-size-phi.ll [hunk 1]
```diff
@@ -61,3 +61,59 @@ if.end:
   %size = call i64 @llvm.objectsize.i64.p0(ptr %p, i1 true, i1 true, i1 false)
   ret i64 %size
 }
+
+define dso_local i64 @pick_max_same(i32 noundef %n) local_unnamed_addr {
```
Comment (nikic) [inline llvm/test/Transforms/LowerConstantIntrinsics/builtin-object-size-phi.ll:65]: Drop dso_local, noundef, local_unnamed_addr, they should not be relevant.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by nikic** [comment]
This change caused a large compile-time regression. I've mostly mitigated this with https://github.com/llvm/llvm-project/commit/c0a64ec13916e420831cee96b48d3ee938f41a66, but there is still some residual regression: http://llvm-compile-time-tracker.com/compare.php?from=0a692b6b9632e1460f9e0e983196f2be5879acd1&to=0bf8763781fa68fa63ee8c1f0d9f6040df97483c&stat=instructions%3Au

**Comment by bevin-hansson** [comment]
That's pretty unfortunate. I'm not sure what more can be done about it, it's bound to iterate further since it's given the chance to.

**Comment by nikic** [comment]
getObjectSize() is almost always called on root instructions (identified objects like allocas, globals) -- actually using it with objectsize intrinsics is rare. I expect this is the additional overhead of the map, not the extra iteration.

**Comment by bevin-hansson** [comment]
Aha, alright. I suppose it makes sense, the tests are probably not built with something like _FORTIFY_SOURCE that would use the intrinsic.

Is the remaining regression acceptable?

**Pull request merged**